server/schedule/region_splitter.go

```diff
@@ -0,0 +1,123 @@
+	cluster opt.Cluster
+	handler SplitRegionsHandler
+}
+
+// NewRegionSplitter return a region splitter
+func NewRegionSplitter(cluster opt.Cluster, handler SplitRegionsHandler) *RegionSplitter {
+	return &RegionSplitter{
+		cluster: cluster,
+		handler: handler,
+	}
+}
+
+// SplitRegions support splitRegions by given split keys.
+func (r *RegionSplitter) SplitRegions(splitKeys [][]byte, retryLimit int) (int, []uint64) {
+	if len(splitKeys) < 1 {
+		return 0, nil
+	}
+	unprocessedKeys := splitKeys
+	newRegions := make(map[uint64]struct{}, len(splitKeys))
+	for i := 0; i <= retryLimit; i++ {
+		if i > 0 {
+			//TODO: sleep for a while
+			time.Sleep(500 * time.Millisecond)
+		}
+		unprocessedKeys = r.splitRegionsByKeys(unprocessedKeys, newRegions)
+		if len(unprocessedKeys) < 1 {
+			break
+		}
+	}
+	returned := make([]uint64, 0, len(newRegions))
+	for regionID := range newRegions {
+		returned = append(returned, regionID)
+	}
+	return 100 - len(unprocessedKeys)*100/len(splitKeys), returned
+}
+
+func (r *RegionSplitter) splitRegionsByKeys(splitKeys [][]byte, newRegions map[uint64]struct{}) [][]byte {
+	//TODO: support batch limit
+	groupKeys, unProcessedKeys := r.groupKeysByRegion(splitKeys)
+	for regionID, keys := range groupKeys {
+		region := r.cluster.GetRegion(regionID)
+		// TODO: assert region is not nil
+		// TODO: assert leader exists
+		// TODO: assert region replicated
+		// TODO: assert region not hot
+		err := r.handler.SplitRegionByKeys(region, keys)
+		if err != nil {
+			unProcessedKeys = append(unProcessedKeys, keys...)
+			continue
+		}
+		// TODO: use goroutine to run watchRegionsByKeyRange asynchronously
+		// TODO: support configure timeout and interval
+		splittedRegionsID := r.handler.WatchRegionsByKeyRange(region.GetStartKey(), region.GetEndKey(), time.Minute, 100*time.Millisecond)
+		for _, id := range splittedRegionsID {
+			newRegions[id] = struct{}{}
+		}
+	}
+	return unProcessedKeys
+}
+
+// GroupKeysByRegion separates keys into groups by their belonging Regions.
+// If any key failed to be found its Region, it will be placed into unProcessed key.
+// If the key is exactly the start key of its region, the key would be discarded directly.
+func (r *RegionSplitter) groupKeysByRegion(keys [][]byte) (map[uint64][][]byte, [][]byte) {
+	unProcessedKeys := make([][]byte, 0, len(keys))
+	groupKeys := make(map[uint64][][]byte, len(keys))
+	for _, key := range keys {
+		region := r.cluster.GetRegionByKey(key)
+		if region == nil {
+			log.Info("region hollow", logutil.ZapRedactByteString("key", key))
+			unProcessedKeys = append(unProcessedKeys, key)
+			continue
+		}
+		// filter start key
+		if bytes.Equal(region.GetStartKey(), key) {
+			continue
+		}
+		group, ok := groupKeys[region.GetID()]
+		if !ok {
+			groupKeys[region.GetID()] = [][]byte{}
+		}
+		log.Info("found region",
+			zap.Uint64("regionID", region.GetID()),
+			logutil.ZapRedactByteString("key", key))
+		groupKeys[region.GetID()] = append(group, key)
```

Comment:
I think we can `groupKeys[region.GetID()] = append(groupKeys[region.GetID()] , key)` directly

Reply:
updated.